staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/library/cost.go

```diff
@@ -0,0 +1,250 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package library
+
+import (
+	"math"
+
+	"github.com/google/cel-go/checker"
+	"github.com/google/cel-go/common"
+	"github.com/google/cel-go/common/types"
+	"github.com/google/cel-go/common/types/ref"
+	"github.com/google/cel-go/common/types/traits"
+	exprpb "google.golang.org/genproto/googleapis/api/expr/v1alpha1"
+)
+
+// CostEstimator implements CEL's interpretable.ActualCostEstimator and checker.CostEstimator.
+type CostEstimator struct {
+	// SizeEstimator provides a CostEstimator.EstimateSize that this CostEstimator will delegate size estimation
+	// calculations to if the size is not well known (i.e. a constant).
+	SizeEstimator checker.CostEstimator
+}
+
+func (l *CostEstimator) CallCost(function, overloadId string, args []ref.Val, result ref.Val) *uint64 {
+	switch function {
+	case "isSorted", "sum", "max", "min", "indexOf", "lastIndexOf":
+		var cost uint64
+		if lister, ok := args[0].(traits.Lister); ok {
+			// Start with a base 1 cost for comparing each element in the list.
+			size := uint64(lister.Size().(types.Int))
+			cost = size
+
+			// If the list contains strings or bytes, add an estimate of the cost of the strings/bytes comparisons
+			// to the base cost.
+			// TODO: This could be optimized by sampling the list instead of traversing it.
+			for it := lister.Iterator(); it.HasNext() == types.True; {
+				v := it.Next()
+				var bytesCompared uint64
+				switch vt := v.(type) {
+				case types.String:
+					bytesCompared += uint64(len(string(vt)))
+				case types.Bytes:
+					bytesCompared += uint64(len([]byte(vt)))
+				default:
+					break // not a string or bytes, so we don't need to traverse the whole list
```

Comment:
would a literal list containing a variable hit this case? does this mean we'll only estimate cost for literal lists containing all literal strings?

Reply:
I've generalized this more and moved it into its own function.